### ISSUE TITLE
docker.Image uses wrong repoDigest when there are multiple

### ISSUE THREAD

**Issue by emil-e** [pulumi/pulumi-docker#778]
### What happened?

I have two stacks, a `test` stack and a `prod` stack, each with their own container registry. When I deploy to both stacks, the `test` stack fails because it picks up the `repoDigest` from the `prod` stack whose container registry it doesn't have access to. 

To illustrate, here is the (redacted) output of `docker inspect`:
```
➜  infra git:(production) ✗ docker inspect mycontainerstest.azurecr.io/test-service:latest
[
    {
        "Id": "sha256:1c16fb012eea50b121496db48500758dab4dcabcccf0b8a13800bbf1ee3f735d",
        "RepoTags": [
            "mycontainersprod.azurecr.io/prod-service:latest",
            "mycontainerstest.azurecr.io/test-service:latest"
        ],
        "RepoDigests": [
            "mycontainersprod.azurecr.io/prod-service@sha256:620312a205fcd362584d80863a1e452b02a5c38d1cfe569523abca59979093df",
            "mycontainerstest.azurecr.io/test-service@sha256:620312a205fcd362584d80863a1e452b02a5c38d1cfe569523abca59979093df"
        ],
    },
    ...
]
```

This is caused by the following code:
https://github.com/pulumi/pulumi-docker/blob/9edb70978358404353b761433e5a34ee5ab5eb44/provider/image.go#L390

As can be seen, it simply picks the first element which is not what one would expect and will cause problems in a case like the one illustrated. Ideally, it would pick the correct one based on the image name and login server.

### Example

I believe the information above should be enough to understand the issue without a complex setup to illustrate the problem

### Output of `pulumi about`

```
CLI          
Version      3.84.0
Go Version   go1.21.1
Go Compiler  gc

Plugins
NAME                      VERSION
azure-native              2.8.0
azuread                   5.42.0
docker                    4.4.1
kubernetes                4.2.0
kubernetes                3.30.2
kubernetes                3.30.2
kubernetes-cert-manager   0.0.5
kubernetes-ingress-nginx  0.0.10
nodejs                    unknown
random                    4.14.0
tls                       4.11.0

Host     
OS       darwin
Version  12.1
Arch     arm64

This project is written in nodejs: executable='/Users/emilsahlen/.nvm/versions/node/v14.17.6/bin/node' version='v14.17.6'

Found no pending operations associated with test

Backend        
Name           Emils-MBP.lan
URL            ___
User           emilsahlen
Organizations  

Dependencies:
NAME                              VERSION
@pulumi/kubernetes                4.2.0
@pulumi/tls                       4.11.0
@pulumi/kubernetes-ingress-nginx  0.0.10
@pulumi/azuread                   5.42.0
@pulumi/docker                    4.4.1
@pulumi/kubernetes-cert-manager   0.0.5
@pulumi/pulumi                    3.83.0
@pulumi/random                    4.14.0
@types/node                       16.18.52
@pulumi/azure-native              2.8.0

Pulumi locates its logs in /var/folders/gh/xrzw20rj3pj5k9hv18x2pzbw0000gn/T/ by default
```

### Contributing

Vote on this issue by adding a 👍 reaction. 
To contribute a fix for this issue, leave a comment (and link to your pull request, if you've opened one already).

**Comment by mikhailshilkov** [comment]
Hi @emil-e Thank you for raising this with us. Could you please provide a code snippet for me to reproduce this problem locally? Ideally, including the setup of Azure Container Registry. Thank you in advance!

**Comment by emil-e** [comment]
```TypeScript
import * as pulumi from "@pulumi/pulumi";
import * as resources from "@pulumi/azure-native/resources";
import * as storage from "@pulumi/azure-native/storage";
import * as containerregistry from "@pulumi/azure-native/containerregistry";
import * as docker from "@pulumi/docker";

const stack = pulumi.getStack();

const resourceGroup = new resources.ResourceGroup(`${stack}-resourcegroup`);

const registry = new containerregistry.Registry(`${stack}-registry`, {
  resourceGroupName: resourceGroup.name,
  registryName: `imagesforpulumibug${stack}`,
  sku: {
    name: containerregistry.SkuName.Basic,
  },
  adminUserEnabled: true,
});

const registryCredentials = containerregistry.listRegistryCredentialsOutput({
  resourceGroupName: resourceGroup.name,
  registryName: registry.name,
});

const image = new docker.Image(`${stack}-image`, {
  imageName: pulumi.interpolate`${registry.loginServer}/${stack}-myimage:latest`,
  registry: {
    server: registry.loginServer,
    username: registryCredentials.username?.apply((u) => u!),
    password: registryCredentials.passwords?.apply((ps) => ps![0].value!),
  },
  build: {
    dockerfile: "Dockerfile",
    context: ".",
    platform: "linux/amd64",
  },
});

export const repoDigest = image.repoDigest;
```

This is the Dockerfile:
```Dockerfile
# syntax=docker/dockerfile:1
FROM scratch
COPY index.ts /
CMD ["echo", "hello"]
```

When I first run `pulumi up` for `stack1`, I get the following output:
```
Outputs:
  + repoDigest: "imagesforpulumibugstack1.azurecr.io/stack1-myimage@sha256:ab619c5f9aab3f416064bbcce02350e87049b8379f6bdccfb883ff225ebacc22"
```

When I then do the same thing for `stack2`, I expect the output `repoDigest` to refer to an image in the container registry of `stack2` but instead it refers to the `repoDigest` of `stack1`. This is because they use the same image with the same ID. It has two `repoDigest`s according to `docker inspect`:
```
[
    {
        "Id": "sha256:63ec01a89dcde35e6499e39ab8c4accccd68198f9284921b666477bc48dd6998",
        "RepoTags": [
            "imagesforpulumibugstack1.azurecr.io/stack1-myimage:latest",
            "imagesforpulumibugstack2.azurecr.io/stack2-myimage:latest"
        ],
        "RepoDigests": [
            "imagesforpulumibugstack1.azurecr.io/stack1-myimage@sha256:ab619c5f9aab3f416064bbcce02350e87049b8379f6bdccfb883ff225ebacc22",
            "imagesforpulumibugstack2.azurecr.io/stack2-myimage@sha256:ab619c5f9aab3f416064bbcce02350e87049b8379f6bdccfb883ff225ebacc22"
        ],
......
```

Pulumi should pick the appropriate one instead of just picking the first one, which is incorrect.

**Comment by mikhailshilkov** [comment]
Thank you for the repro @emil-e, I can see the same issue locally. We'll take a look at fixing it. Let me know if you want to try to open a PR for it, since you already dove into the code.

**Comment by emil-e** [comment]
> Thank you for the repro @emil-e, I can see the same issue locally. We'll take a look at fixing it. Let me know if you want to try to open a PR for it, since you already dove into the code.

I could maybe do that if you could point me to some docs on how to run Pulumi with `pulumi-docker` "from source". I didn't find much info about what's a good "debugging workflow".

**Comment by guineveresaenger** [comment]
Hi everyone - we have pushed a fix for this as part of #843. We believe it addresses this issue and have a [beta release](https://github.com/pulumi/pulumi-docker/releases/tag/v4.5.0-beta.1) out for you to try. We anticipate rolling this into the next stable release within a few days.
